### ISSUE TITLE
Dart unittests can't import mojom-generated code.

### ISSUE THREAD

**Issue by ppiet** [domokit/mojo#618]
A dart unittest cannot import mojom-generated bindings, otherwise this happens:

```
  The built-in library 'dart:mojo.internal' is not available on the stand-alone VM.
  'package:mojo/core.dart': error: line 10 pos 1: library handler failed
  import 'dart:mojo.internal';
  ^
```

and reportedly just commenting out 'import dart:mojo.internal'  in 'mojo/core.dart' makes things work. It would be nice to allow unit-testing against mojom-generated datatypes, so that developers coming from the dart world can write tests that are familiar to them.

@johnmccutchan

**Comment by zanderso** [comment]
Unless this is still a big impediment to testing, I'm not sure that this should be a high priority. 

@jimbeveridge @tonygentilcore How is it going?

**Comment by jimbeveridge** [comment]
I agree this isn't a P0 or P1, but IMO it's a developer experience issues that needs to be addressed.

**Comment by floitschG** [comment]
We should be able to fix this once configuration-specific imports are implemented (dart-lang/sdk#24581)

**Comment by zanderso** [comment]
Yah, I would be in favor of addressing this with configuration-specific imports rather than attempting a refactoring of the Dart bindings to simulate it.

**Comment by johnmccutchan** [comment]
I'm not even sure the original request is valid. Dart programs that import 'dart:html' do not run under the standalone embedder. I'm not sure how 'dart:mojo.internal' or 'dart:ui' are any different.

**Comment by jimbeveridge** [comment]
In my experience so far, the unit tests don't actually make any calls on the service proxy, nor do they run the message loop, so the code in dart:mojo.internal is not needed.

For dart:ui, it seems like it should work if you are running headless and nothing is actually being rendered (I'm not sure if it does work at the moment, but it would be nice if it did.)

I agree with zra about not refactoring the bindings. My changes to the bindings were a workaround and not a path I recommend for moving forward.

**Comment by pylaligand** [comment]
It looks like config-specific imports are now available - though currently hidden behind a flag. Any chance we can unblock the present issue? This is getting in the way of more projects.
(happy to help btw)

**Comment by floitschG** [comment]
Normally we have to go through a (sometimes lengthy) process of approving new language features.
I just sent a mail, trying to take some shortcuts and get the approval faster. Hopefully I will have news within the next few days.

**Comment by floitschG** [comment]
CL is up: https://chromiumcodereview.appspot.com/1851753002/

**Comment by zanderso** [comment]
@floitschG is there any update on this? Thanks!

**Comment by floitschG** [comment]
Yes. Thanks for reminding.
Configuration-specific imports are now (silently) enabled. If Mojo uses a recent VM snapshot, we should be able to modify the Mojo package to run on the VM. As a first step I would just stub out the functions and make them throw. This should allow tests that don't depend on Mojo to run (even if they transitively import the mojo package).

I would prefer if we added a 'dart:mojo' first (so that the config-import is guarded by a check against `dart.library.mojo` instead of `dart.library.mojo.internal`), but we can change that at a later point.

**Comment by zanderso** [comment]
I rolled the VM in Mojo to https://github.com/dart-lang/sdk/commit/e4db01fdd43f019988a901eb51c72790652760a4 last week. Is that recent enough? Is it under a flag?

By adding `dart:mojo`, do you mean renaming `dart:mojo.internal` to `dart:mojo`? I think that should be easy enough to do, but I kind of like having the `.internal` there to warn folks off trying to use it. If you don't like `.internal`, how about `dart:mojo.system`?

**Comment by floitschG** [comment]
That should be recent enough (it was in 1.16).

I wouldn't actually rename `dart:mojo.internal`, but rather add an empty `dart:mojo`, potentially having a dartdoc that explains that users should use the mojo package instead.
I guess, I would rename `dart:mojo.internal` to `dart:mojo._internal` to make clear that this is a private library, but "internal" pretty much implies that already, so it's not that important.

**Comment by zanderso** [comment]
@floitschG's work has landed in the VM and in the mojo package. There is a simple test in the Mojo package here:

https://github.com/domokit/mojo/blob/master/mojo/dart/packages/mojo/test/standalone_test.dart

demonstrating an import of package:mojo/core.dart (which imports dart:mojo.internal) that runs in the standalone VM. Imports of generated bindings will work, but anything that relies on an actual Mojo system being attached will throw an exception. Hopefully this is not a problem for unit testing, though.

Please give this a try and file a new issue if you run into any problems. Thanks!
